File: leetcode/Stack/0155_Min_Stack.py
class MinStack(object):

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)
        self.stack.append(val)

    def pop(self):
        """
        :rtype: None
        """
        last_num = self.stack.pop()
        if last_num == self.min_stack[-1]:
            self.min_stack.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.min_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: leetcode/Stack/0155_Min_Stack.cpp
#include <stack>
class MinStack {
public:
    void push(int x) {
        if (minStack.empty() || x <= minStack.top())
            minStack.push(x);
        Stack.push(x);
    }

    void pop() {
        if (Stack.top() == minStack.top())
            minStack.pop();
        Stack.pop();
    }

    int top() {
        return Stack.top();
    }

    int getMin() {
        return minStack.top();
    }
private:
    stack<int> Stack;
    stack<int> minStack;
};